### ISSUE TITLE
research: explore to provide a theme for flatpickr

### ISSUE THREAD

**Issue by mariohamann** [solid-design-system/solid#516]
# User story

As a developer I want to use a datepicker in the design of solid design system.

# Description

https://flatpickr.js.org/ is a massive library which supports all bigger JS frameworks and supports themes. We should explore if it makes sense to just create a theme for that and people can implement flatpickr theirself.

**Comment by abudd1094** [comment]
**Findings:**

custom flatpickr styles must be implemented on a per class basis via CSS.  I started from one of the default themes: dark.css.  The initial CSS files are over 800 lines and not trivial or self-explanatory.  Starting from scratch is also not viable, so one must be careful to adapt the existing defaults to a custom design.

Solid Design:
![image](https://github.com/solid-design-system/solid/assets/39494579/3ac78653-a6cc-4bf0-a95f-50ad46becf45)

Solid flatpckr Demo:
![image](https://github.com/solid-design-system/solid/assets/39494579/5c4a9860-e6f2-4c2f-b3fe-05ab9b171b35)

The current Figma design for a datepicker component does not make sense anymore as this is only a CSS style sheet for overriding the flatpickr styles.  Furthermore, the config options for flatpickr are more limited than a purely custom implementation.  

There is no opportunity for 'slotting' within the flatpickr:
![image](https://github.com/solid-design-system/solid/assets/39494579/b01632b7-6d87-4412-913a-7c0a2c6f0409)

flatpickr offers a standard select menu for selecting a month.  The current design will not work:

Solid Design:
![image](https://github.com/solid-design-system/solid/assets/39494579/eab63127-54ab-47d1-988f-9fac9921c962)

Solid flatpickr Demo:
![image](https://github.com/solid-design-system/solid/assets/39494579/ae2b981e-6fca-4e8e-abf4-e55f20635de5)

Likewise, the year is adjusted using a stepper.  The current design will not work:

Solid Design:
![image](https://github.com/solid-design-system/solid/assets/39494579/034fc011-576a-4541-81cd-dcd98f5abbd7)

Solid flatpickr Demo:
![image](https://github.com/solid-design-system/solid/assets/39494579/f73c1d7f-7715-4b24-884a-ebcc5e84dae3)

**Comment by mariohamann** [comment]
For month view: https://flatpickr.js.org/plugins/ Does it work with the plugins?

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by abudd1094** [comment]
@mariohamann yes there is a month view plugin.  But if design is expecting the month view to show up when clicking the month, that is not how flatpickr is designed.  It will have to be manipulated.

I'll try the month view plugin in the morning, the rangePlugin works, and is on my branch.

**Comment by karlbaumhauer** [comment]
thx at @abudd1094 for the research. 
In order to put a higher priority on the performance and building up a lightweight library, I would highly suggest to adopt the design towards the possibilities flatpickr offers. 

As I saw, there has already been set up a meeting to evaluate, I am happy to get an update afterwards.

**Comment by abudd1094** [comment]
after initial meeting on 15.11.2023 I am investigating the accessibility checklist further and will leave another findings report on this ticket

**Comment by abudd1094** [comment]
I have addressed each bullet point from my previous comment in detail below:

# Findings

[POC branch](https://github.com/solid-design-system/solid/tree/poc/flatpickr): contains dummy component called ’sd-datepicker’ which implements an instance of flatpickr for demonstration purposes
git checkout poc/flatpickr

**Updating date selection on calendar UI when updating text input manually**

This can be achieved by setting the flatpickr config option ‘allowInput’ to ‘true’.
Then we can implement an input listener that checks for a valid date and uses several flatpickr instance methods to instantly trigger UI changes.  In this case the selectedDates array is set to the validated input date and then the jumpToDate function brings the UI to the correct year and month.

**Date picker is fully accessible via keyboard**

This is possible.  The default implementation yielded strange behaviors and did not allow for freely typing in the input element.  If I configure ‘allowInput’ to be ‘true’, I can add a keydown handler to focus on the first visible day (effectively shifting keyboard focus to the calendar UI) when pressing on the down arrow key.

**Date picker is compatible with screen readers**

I have limited experience with screen readers.  I investigated here by enabling a simple [Google Chrome extension](https://chromewebstore.google.com/detail/screen-reader/kgejglhpjiefppelpmljglcjbhoiplfn?pli=1).  The result was clearly understandable dates are read aloud as the user keys over the various dates.  My conclusion, compatibility seems to have been considered and most likely functions because of semantic use of HTML in the Flatpickr implementation.

**Possibility to display two calendars side by side for multi month selections**

This can be achieved by setting the flatpickr config option ’showMonths’ to 2 or higher.  The ‘mode’ config option can additionally be set to ‘multiple' or 'range’ which is the most likely use case.  In sd-datepicker, this yielded a crunched and visually unpleasing result, however [this code pen](https://codepen.io/mtmtkzm/pen/OEQvLB) shows the concept working well.  I imagine with additional CSS investigation we can get it working in our contexts.

**Possibility to use custom Month and Year picker UI’s**

This can also be achieved.  In the first instance we can use the monthSelectionPlugin, however this is not easily achieved in combination with the standard calendar date picker.  I experimented with creating my own plugin which essentially uses the documented hooks and events to mutate the date picker as we see fit.  For this experiment, I started with the monthSelectPlugin to see if I can alter the DOM in such a way that would let us implement a month picker UI that appears when the user clicks on the ‘monthDropdown’.  However, it gets very complicated very quickly as we regress to DOM level programming in undocumented territory.  

Despite this, I took a look at the existing plugins and created a file called ‘pluginTemplate.js’ which I believe can serve as a good starting point if we wish to author our own plugin.  It is important to note that authoring custom plugins is not discussed in the flatpickr docs or more broadly online.  This says to me that it has not reached a maturity level where custom plugins will be consistently supported and furthermore only some of the available hooks are documented on the official site.  Regardless, I left a file that works on our context with commentary explaining each part up to my current understanding level.

# Concerns
Custom plugins: discussed above...
Custom themes: themes are likewise not extensively documented.  There are several provided but no information on customizing.  We can simply use the classes like I have in solid.css but I would be cautious about future updates mutating classNames or something detrimental along those lines.
Extent of UI influence: Both of the above concerns indicate that we may have to compromise on design (e.g. with the month and year pickers).

# Conclusion
My ultimate conclusion is that flatpickr is a mature enough solution to be used however we should be cautious when adapting it for our context.

# Additional Questions
Simply out of curiosity as I imagine this was previously discussed.

Why not adapt the old Union date picker?  Seems like a lot of work went into it.

**Comment by karlbaumhauer** [comment]
Awesome. Thx a lot for the detailed report @abudd1094 

What was the agreement on how we proceed here now?

@mariohamann / @coraliefeil should we have a closer look into a possible datepicker MVP in the next pre-refinement? I would open up a ticket for it.

**Comment by abudd1094** [comment]
Next pre refinement sounds right to me, we didn't have an explicit agreement other than I'd follow up here.  I would be happy to try and implement this for the library if / when the time comes.  Let me know if you need any further info or input from my side.

**Comment by coraliefeil** [comment]
Thank you for the fantastic research and also for finding the codepen example for multiple calenders. 

Will ask around if more than 1 team needs that feature.

**Comment by mariohamann** [comment]
@abudd1094 Thanks for the research!

First of all: I protected your branch so that no one deletes it per accident.

I see two main issues for integrating an own component:
## Bundle size

I just created a PR to see how much the bundle size would increase:

<img width="758" alt="CleanShot 2023-11-20 at 12 09 39@2x" src="https://github.com/solid-design-system/solid/assets/26542182/dec4255f-62bc-4d63-ba0f-b5ddaa16a625">

Such an increase is way too much for an component which is barely in use.

## Initial costs
Outlying an appropriate API either hard work or we just allow to let people set a config as done in Vue: https://github.com/ankurk91/vue-flatpickr-component

I don't see much benefit here if we provide it. On top we would have to build a lot of stuff because of ShadowDOM. If we had the sd-inputs inside our component, we would have to do a bunch of stuff to make it work with forms. If we leave it in the LightDOM via slotting, I'm not sure if we will get it working with FlatPickr.

When re-using the datepicker component Oezlem created a bunch of features are missing, like multiple month view, selecting ranges etc.

## Maintenance
Using such an enormous dependency will force us to maintain bugs, even if they are in flatpickr. As you already outlined, the repo doesn't seem to be very active.

---

## Conclusion

That said, I wouldn't exclude shipping an own component FOREVER. But at the moment, I don't see this worth facing our current and future resources.

**Comment by mariohamann** [comment]
@coraliefeil @abudd1094 @karlbaumhauer 

From my perspective we can close this ticket an can refine the syn-datepicker. There we should handle following topics:
1. syn-flatpickr CSS Style
2. Code splitting for CSS styles and omitting syn-flatpickr from style bundle
3. Stories which show most important functionality.
4. We should think about how to provide the accessibility feature @abudd1094 mentioned, e. g. if we provide a JS file containing the hook which we will redistribute (but in which package?)

**Comment by abudd1094** [comment]
What is "syn-flatpickr" or "syn-datepicker" ? @mariohamann

**Comment by mariohamann** [comment]
@abudd1094 Sorry, that was the prefix of a another design system. 😆
